tools/hsmtool.c

```diff
@@ -368,6 +371,91 @@ static int guess_to_remote(const char *address, struct node_id *node_id,
 	return 1;
 }
 
+static void read_mnemonic(char *mnemonic) {
+	printf("Introduce the BIP39 24 word list:\n");
+	char word[50];
+	uint n_words;
+	for (n_words = 0; n_words < 24; n_words++) {
+		scanf("%s", word);
+		/* We need to concat the strings using ' ' so libwally can
+		 * validate it */
+		if (n_words == 0) {
+			strcpy(mnemonic, word);
+		} else {
+			strcat(mnemonic, " ");
+			strcat(mnemonic, word);
+		}
+	}
+	/* scanf() leaves the newline character in the input buffer */
+	getchar();
+}
+
+static void read_passphrase(char *passphrase) {
+	struct termios current_term, temp_term;
+	printf("Warning: remember that different passphrase yield different "
+	       "bitcoin wallets.\n");
+	printf("Leave it empty for using the value by default (echo is "
+	       "disabled now).\n");
+	printf("Enter your passphrase: ");
+
+	/* Change terminal options so we do not echo the passphrase */
+	if (tcgetattr(fileno(stdin), &current_term) != 0)
+		errx(ERROR_USAGE, "Could not get current terminal options.");
+	temp_term = current_term;
+	temp_term.c_lflag &= ~ECHO;
+	if (tcsetattr(fileno(stdin), TCSAFLUSH, &temp_term) != 0)
+		errx(ERROR_USAGE, "Could not disable passphrase echoing.");
+	/* If we don't flush we might end up being buffered and we might seem
+	 * to hang while we wait for the password. */
+	fflush(stdout);
+
+	size_t passphrase_size;
+	if (getline(&passphrase, &passphrase_size, stdin) < 0)
+		errx(ERROR_USAGE, "Could not read passphrase from stdin.");
+	printf("passphrase size: %ld\n", passphrase_size);
+	/* Replace the last character which will always be '\n' */
+	passphrase[strlen(passphrase) - 1] = '\0';
+
+	if (tcsetattr(fileno(stdin), TCSAFLUSH, &current_term) != 0)
+		errx(ERROR_USAGE, "Could not restore terminal options.");
+}
+
+static int generate_hsm(const char *hsm_secret_path)
+{
+	char mnemonic[BIP39_WORDLIST_LEN];
+	read_mnemonic(mnemonic);
+	if (bip39_mnemonic_validate(NULL, mnemonic) != 0)
```

Comment:
does this correctly validate non-English word list sets? You might try the test cases from https://github.com/bip32JP/bip32JP.github.io/blob/master/test_JP_BIP39.json